k8s-scheduler/src/test/java/com/vmware/dcm/ScopeTest.java

```diff
@@ -0,0 +1,280 @@
+        final PodResourceEventHandler handler = new PodResourceEventHandler(eventHandler::handle);
+
+        final int numNodes = 100;
+        final int numPods = 40;
+
+        for (int i = 0; i < numNodes; i++) {
+            final Node node = newNode("n" + i, Collections.emptyMap(), Collections.emptyList());
+            // nodes have more spare capacity as i increases
+            node.getStatus().getCapacity().put("cpu", new Quantity(String.valueOf(10)));
+            node.getStatus().getCapacity().put("memory", new Quantity(String.valueOf(1000 + (1000 * i) / numNodes)));
+            node.getStatus().getCapacity().put("pods", new Quantity(String.valueOf(100)));
+            nodeResourceEventHandler.onAddSync(node);
+
+            // Add one system pod per node
+            final String podName = "system-pod-n" + i;
+            final Pod pod = newPod(podName, "Running");
+            pod.getSpec().setNodeName("n" + i);
+            handler.onAddSync(pod);
+        }
+
+        for (int i = 0; i < numPods; i++) {
+            handler.onAddSync(newPod("p" + i));
+        }
+
+        final Scheduler scheduler = new Scheduler(dbConnectionPool, policies, "ORTOOLS", true, NUM_THREADS);
+        scheduler.setScopeOn();
+        scheduler.scheduleAllPendingPods(new EmulatedPodToNodeBinder(dbConnectionPool));
+
+        // Check that all pods have been scheduled to a node eligible by the scope filtering
+        final Result<PodInfoRecord> fetch = conn.selectFrom(Tables.POD_INFO).fetch();
+        assertEquals(numNodes + numPods, fetch.size());
+        fetch.forEach(e -> assertTrue(e.getNodeName() != null
+                && e.getNodeName().startsWith("n")
+                // check that new pods have been scheduled to the last, least loaded nodes
+                && (Integer.parseInt(e.getNodeName().substring(1)) >= numNodes - numPods
+                || !e.getPodName().startsWith("p"))));
+    }
+
+    /*
+     * E2E test with scheduler:
+     * Test if Scope filters out nodes maintaining while maintaining matching labels.
+     * Labeled nodes have low spare capacity.
+     */
+    @Test
+    public void testScopedSchedulerNodeLabels() {
+        final DBConnectionPool dbConnectionPool = new DBConnectionPool();
+        final DSLContext conn = dbConnectionPool.getConnectionToDb();
+        final List<String> policies = Policies.getDefaultPolicies();
+        final NodeResourceEventHandler nodeResourceEventHandler = new NodeResourceEventHandler(dbConnectionPool);
+
+        final PodEventsToDatabase eventHandler = new PodEventsToDatabase(dbConnectionPool);
+        final PodResourceEventHandler handler = new PodResourceEventHandler(eventHandler::handle);
+
+        final int numNodes = 1000;
+        final int numPods = 20;
+
+        final List<Integer> gpuNodesIdx = List.of(0, 1);
+        final List<Integer> ssdNodesIdx = List.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12);
```

Comment:
Is there a risk that picking earlier IDs for pods/nodes biases the test somehow (for example, if the underlying selects are by chance, likely to yield earlier rows than later ones)?